`make stress-test-e2e` run detects a race condition in test framework.go

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via 
-->
Golang stress tests detects a race condition.

**What happened**:
Data race detected on 7 minute of stress E2E test run.
```
      /usr/local/go/src/testing/testing.go:672 +0x9e                                                                                                                                                     
  agones.dev/agones/test/e2e/framework.(*Framework).WaitForFleetCondition()                                                                                                                              
      /go/src/agones.dev/agones/test/e2e/framework/framework.go:146 +0x504
  agones.dev/agones/test/e2e.scaleAndWait()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:1005 +0xc4
  agones.dev/agones/test/e2e.TestScaleUpAndDownInParallelStressTest.func1()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:824 +0xe4
```
**What you expected to happen**:
No race conditions.

**How to reproduce it (as minimally and precisely as possible)**:
```
make gcloud-test-cluster
STRESS_TEST_LEVEL=4 make stress-test-e2e
```

**Anything else we need to know?**:

**Environment**:
- Agones version: 0.12
- Kubernetes version (use `kubectl version`): 1.12
- Cloud provider or hardware configuration: Google Cloud
- Install method (yaml/helm): yaml
- Troubleshooting guide log(s): 

```
time="2019-09-10 19:29:37.352" level=info msg="fleet scale-fleet-7-b8l2k has 26/40 ready replicas"                                                                                                       
time="2019-09-10 19:29:37.353" level=info msg="error waiting for fleet condition" error="timed out waiting for the condition" fleet=scale-fleet-7-b8l2k                                                  
==================                                                                                                                                                                                       
WARNING: DATA RACE                                                                                                                                                                                       
Write at 0x00c0002fe151 by goroutine 86:                                                                                                                                                                 
  testing.(*common).FailNow()                                                                                                                                                                            
      /usr/local/go/src/testing/testing.go:608 +0x4c                                                                                                                                                     
  testing.(*common).Fatalf()
      /usr/local/go/src/testing/testing.go:672 +0x9e                                                                                                                                                     
  agones.dev/agones/test/e2e/framework.(*Framework).WaitForFleetCondition()
      /go/src/agones.dev/agones/test/e2e/framework/framework.go:146 +0x504
  agones.dev/agones/test/e2e.scaleAndWait()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:1005 +0xc4
  agones.dev/agones/test/e2e.TestScaleUpAndDownInParallelStressTest.func1()                                                                                                                              
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:824 +0xe4

Previous write at 0x00c0002fe151 by goroutine 82:
  testing.(*common).FailNow()                                                                                                                                                                            
      /usr/local/go/src/testing/testing.go:608 +0x4c
  testing.(*common).Fatalf()
      /usr/local/go/src/testing/testing.go:672 +0x9e                                                                                                                                                     
  agones.dev/agones/test/e2e/framework.(*Framework).WaitForFleetCondition()                                                                                                                              
      /go/src/agones.dev/agones/test/e2e/framework/framework.go:146 +0x504
  agones.dev/agones/test/e2e.scaleAndWait()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:1005 +0xc4
  agones.dev/agones/test/e2e.TestScaleUpAndDownInParallelStressTest.func1()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:824 +0xe4

Goroutine 86 (running) created at:
  agones.dev/agones/test/e2e.TestScaleUpAndDownInParallelStressTest()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:809 +0x128d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 82 (finished) created at:
  agones.dev/agones/test/e2e.TestScaleUpAndDownInParallelStressTest()
      /go/src/agones.dev/agones/test/e2e/fleet_test.go:809 +0x128d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163
==================

...
time="2019-09-10 19:29:47.609" level=info msg=stress_4_fleet_40_scale_down avg=6.64017396 count=5 duration=4.7771784 max=9.9468814 min=5.1667113 p50=5.791677825 p90=9.473440700000001 p95=9.71016105 p99=9.899537330000001 p999=9.942146993000001
--- FAIL: TestScaleUpAndDownInParallelStressTest (355.56s)
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-3-jj5fc
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-7-b8l2k
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-1-hkw7f
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-9-82n47
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-5-tht4x
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-8-cbtwh
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-0-zsjtc
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-2-6r72j
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-6-4krtn
    fleet_test.go:1005: error waiting for fleet condition on fleet scale-fleet-4-sx8dp
    testing.go:809: race detected during execution of test
FAIL
time="2019-09-10 19:29:47.617" level=info msg="Cleaning up now."
time="2019-09-10 19:30:08.409" level=info msg="Finished cleanup."
FAIL    agones.dev/agones/test/e2e      378.119s
```
- Others:

--- a/test/e2e/allocator_test.go
+++ b/test/e2e/allocator_test.go
@@ -74,7 +74,7 @@ func TestAllocator(t *testing.T) {
 	if !assert.Nil(t, err) {
 		return
 	}
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 	gsa := &allocationv1.GameServerAllocation{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: namespace,
@@ -169,7 +169,7 @@ func TestAllocatorCrossNamespace(t *testing.T) {
 	if !assert.Nil(t, err) {
 		return
 	}
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 	gsa := &allocationv1.GameServerAllocation{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: namespaceA,

--- a/test/e2e/fleet_test.go
+++ b/test/e2e/fleet_test.go
@@ -71,7 +71,7 @@ func TestFleetScaleUpEditAndScaleDown(t *testing.T) {
 
 			assert.Equal(t, int32(1), flt.Spec.Replicas)
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			// scale up
 			const targetScale = 3
@@ -82,10 +82,10 @@ func TestFleetScaleUpEditAndScaleDown(t *testing.T) {
 				flt = scaleFleetSubresource(t, flt, targetScale)
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
 			gsa := framework.CreateAndApplyAllocation(t, flt)
 
-			framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+			framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 				return fleet.Status.AllocatedReplicas == 1
 			})
 
@@ -126,16 +126,16 @@ func TestFleetScaleUpEditAndScaleDown(t *testing.T) {
 			} else {
 				flt = scaleFleetSubresource(t, flt, scaleDownTarget)
 			}
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(0))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(0))
 
 			// delete the allocated GameServer
 			gp := int64(1)
 			err = client.GameServers(defaultNs).Delete(gsa.Status.GameServerName, &metav1.DeleteOptions{GracePeriodSeconds: &gp})
 			assert.Nil(t, err)
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(1))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(1))
 
-			framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+			framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 				return fleet.Status.AllocatedReplicas == 0
 			})
 		})
@@ -174,7 +174,7 @@ func TestFleetRollingUpdate(t *testing.T) {
 
 				assert.Equal(t, int32(1), flt.Spec.Replicas)
 
-				framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+				framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 				// scale up
 				const targetScale = 8
@@ -185,7 +185,7 @@ func TestFleetRollingUpdate(t *testing.T) {
 					flt = scaleFleetSubresource(t, flt, targetScale)
 				}
 
-				framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
+				framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
 
 				flt, err = client.Fleets(defaultNs).Get(flt.ObjectMeta.GetName(), metav1.GetOptions{})
 				assert.NoError(t, err)
@@ -246,9 +246,9 @@ func TestFleetRollingUpdate(t *testing.T) {
 					flt = scaleFleetSubresource(t, flt, scaleDownTarget)
 				}
 
-				framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(1))
+				framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(1))
 
-				framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+				framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 					return fleet.Status.AllocatedReplicas == 0
 				})
 			})
@@ -277,7 +277,7 @@ func TestScaleFleetUpAndDownWithGameServerAllocation(t *testing.T) {
 
 			assert.Equal(t, int32(1), flt.Spec.Replicas)
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			// scale up
 			const targetScale = 3
@@ -288,7 +288,7 @@ func TestScaleFleetUpAndDownWithGameServerAllocation(t *testing.T) {
 				flt = scaleFleetSubresource(t, flt, targetScale)
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
 
 			// get an allocation
 			gsa := &allocationv1.GameServerAllocation{ObjectMeta: metav1.ObjectMeta{GenerateName: "allocation-"},
@@ -299,7 +299,7 @@ func TestScaleFleetUpAndDownWithGameServerAllocation(t *testing.T) {
 			gsa, err = framework.AgonesClient.AllocationV1().GameServerAllocations(defaultNs).Create(gsa)
 			assert.Nil(t, err)
 			assert.Equal(t, allocationv1.GameServerAllocationAllocated, gsa.Status.State)
-			framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+			framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 				return fleet.Status.AllocatedReplicas == 1
 			})
 
@@ -311,15 +311,15 @@ func TestScaleFleetUpAndDownWithGameServerAllocation(t *testing.T) {
 				flt = scaleFleetSubresource(t, flt, scaleDownTarget)
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(0))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(0))
 
 			// delete the allocated GameServer
 			gp := int64(1)
 			err = client.GameServers(defaultNs).Delete(gsa.Status.GameServerName, &metav1.DeleteOptions{GracePeriodSeconds: &gp})
 			assert.Nil(t, err)
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(1))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(1))
 
-			framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+			framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 				return fleet.Status.AllocatedReplicas == 0
 			})
 		})
@@ -407,7 +407,7 @@ func TestUpdateGameServerConfigurationInFleet(t *testing.T) {
 
 	assert.Equal(t, int32(replicasCount), flt.Spec.Replicas)
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	// get an allocation
 	gsa := &allocationv1.GameServerAllocation{ObjectMeta: metav1.ObjectMeta{GenerateName: "allocation-"},
@@ -418,7 +418,7 @@ func TestUpdateGameServerConfigurationInFleet(t *testing.T) {
 	gsa, err = framework.AgonesClient.AllocationV1().GameServerAllocations(defaultNs).Create(gsa)
 	assert.Nil(t, err, "cloud not create gameserver allocation")
 	assert.Equal(t, allocationv1.GameServerAllocationAllocated, gsa.Status.State)
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.AllocatedReplicas == 1
 	})
 
@@ -454,7 +454,7 @@ func TestReservedGameServerInFleet(t *testing.T) {
 		defer client.Fleets(defaultNs).Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	gsList, err := framework.ListGameServersFromFleet(flt)
 	assert.NoError(t, err)
@@ -468,16 +468,16 @@ func TestReservedGameServerInFleet(t *testing.T) {
 	assert.NoError(t, err)
 
 	// make sure counts are correct
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.ReadyReplicas == 2 && fleet.Status.ReservedReplicas == 1
 	})
 
 	// scale down to 0
 	flt = scaleFleetSubresource(t, flt, 0)
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(0))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(0))
 
 	// one should be left behind
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		result := fleet.Status.ReservedReplicas == 1
 		logrus.WithField("reserved", fleet.Status.ReservedReplicas).WithField("result", result).Info("waiting for 1 reserved replica")
 		return result
@@ -615,7 +615,7 @@ func TestGameServerAllocationDuringGameServerDeletion(t *testing.T) {
 
 		assert.Equal(t, size, flt.Spec.Replicas)
 
-		framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+		framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 		var allocs []string
 
@@ -730,14 +730,14 @@ func TestCreateFleetAndUpdateScaleSubresource(t *testing.T) {
 		defer client.Fleets(defaultNs).Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 	assert.Equal(t, initialReplicas, flt.Spec.Replicas)
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	newReplicas := initialReplicas * 2
 	scaleFleetSubresource(t, flt, newReplicas)
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(newReplicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(newReplicas))
 
 	scaleFleetSubresource(t, flt, initialReplicas)
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(initialReplicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(initialReplicas))
 }
 
 // TestScaleUpAndDownInParallelStressTest creates N fleets, half of which start with replicas=0
@@ -796,13 +796,14 @@ func TestScaleUpAndDownInParallelStressTest(t *testing.T) {
 	// wait for initial fleet conditions.
 	for fleetNumber, flt := range fleets {
 		if fleetNumber%2 == 0 {
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(fleetSize))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(fleetSize))
 		} else {
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(0))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(0))
 		}
 	}
-
+	errors := make(chan error)
 	var wg sync.WaitGroup
+	finished := make(chan bool, 1)
 
 	for fleetNumber, flt := range fleets {
 		wg.Add(1)
@@ -815,19 +816,46 @@ func TestScaleUpAndDownInParallelStressTest(t *testing.T) {
 			}()
 
 			if fleetNumber%2 == 0 {
-				scaleDownStats.ReportDuration(scaleAndWait(t, flt, 0), nil)
+				duration, err := scaleAndWait(t, flt, 0)
+				if err != nil {
+					fmt.Println(err)
+					errors <- err
+					return
+				}
+				scaleDownStats.ReportDuration(duration, nil)
 			}
 			for i := 0; i < repeatCount; i++ {
 				if time.Now().After(deadline) {
 					break
 				}
-				scaleUpStats.ReportDuration(scaleAndWait(t, flt, fleetSize), nil)
-				scaleDownStats.ReportDuration(scaleAndWait(t, flt, 0), nil)
+				duration, err := scaleAndWait(t, flt, fleetSize)
+				if err != nil {
+					fmt.Println(err)
+					errors <- err
+					return
+				}
+				scaleUpStats.ReportDuration(duration, nil)
+				duration, err = scaleAndWait(t, flt, 0)
+				if err != nil {
+					fmt.Println(err)
+					errors <- err
+					return
+				}
+				scaleDownStats.ReportDuration(duration, nil)
 			}
 		}(fleetNumber, flt)
 	}
+	go func() {
+		wg.Wait()
+		close(finished)
+	}()
 
-	wg.Wait()
+	select {
+	case <-finished:
+	case err := <-errors:
+		t.Fatalf("Error in waiting for a fleet to scale: %s", err)
+	}
+	fmt.Println("We are Done")
 }
 
 // Creates a fleet and one GameServer with Packed scheduling.
@@ -851,11 +879,11 @@ func TestUpdateFleetScheduling(t *testing.T) {
 			assert.Equal(t, int32(1), flt.Spec.Replicas)
 			assert.Equal(t, apis.Packed, flt.Spec.Scheduling)
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			const targetScale = 2
 			flt = schedulingFleetPatch(t, flt, apis.Distributed, targetScale)
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(targetScale))
 
 			assert.Equal(t, int32(targetScale), flt.Spec.Replicas)
 			assert.Equal(t, apis.Distributed, flt.Spec.Scheduling)
@@ -929,7 +957,7 @@ func TestFleetRecreateGameServers(t *testing.T) {
 				defer client.Fleets(defaultNs).Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			list, err := listGameServers(flt, client)
 			assert.NoError(t, err)
@@ -953,7 +981,7 @@ func TestFleetRecreateGameServers(t *testing.T) {
 				assert.NoError(t, err)
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 		})
 	}
 }
@@ -999,11 +1027,12 @@ func schedulingFleetPatch(t *testing.T,
 	return fltRes
 }
 
-func scaleAndWait(t *testing.T, flt *agonesv1.Fleet, fleetSize int32) time.Duration {
+func scaleAndWait(t *testing.T, flt *agonesv1.Fleet, fleetSize int32) (duration time.Duration, err error) {
 	t0 := time.Now()
 	scaleFleetSubresource(t, flt, fleetSize)
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(fleetSize))
-	return time.Since(t0)
+	err = framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(fleetSize))
+	duration = time.Since(t0)
+	return
 }
 
 // scaleFleetPatch creates a patch to apply to a Fleet.

--- a/test/e2e/fleetautoscaler_test.go
+++ b/test/e2e/fleetautoscaler_test.go
@@ -52,7 +52,7 @@ func TestAutoscalerBasicFunctions(t *testing.T) {
 		defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	fleetautoscalers := framework.AgonesClient.AutoscalingV1().FleetAutoscalers(defaultNs)
 	fas, err := fleetautoscalers.Create(defaultFleetAutoscaler(flt))
@@ -66,36 +66,36 @@ func TestAutoscalerBasicFunctions(t *testing.T) {
 
 	// the fleet autoscaler should scale the fleet up now up to BufferSize
 	bufferSize := int32(fas.Spec.Policy.Buffer.BufferSize.IntValue())
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
 
 	// patch the autoscaler to increase MinReplicas and watch the fleet scale up
 	fas, err = patchFleetAutoscaler(fas, intstr.FromInt(int(bufferSize)), bufferSize+2, fas.Spec.Policy.Buffer.MaxReplicas)
 	assert.Nil(t, err, "could not patch fleetautoscaler")
 
 	// min replicas is now higher than buffer size, will scale to that level
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(fas.Spec.Policy.Buffer.MinReplicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(fas.Spec.Policy.Buffer.MinReplicas))
 
 	// patch the autoscaler to remove MinReplicas and watch the fleet scale down to bufferSize
 	fas, err = patchFleetAutoscaler(fas, intstr.FromInt(int(bufferSize)), 0, fas.Spec.Policy.Buffer.MaxReplicas)
 	assert.Nil(t, err, "could not patch fleetautoscaler")
 
 	bufferSize = int32(fas.Spec.Policy.Buffer.BufferSize.IntValue())
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
 
 	// do an allocation and watch the fleet scale up
 	gsa := framework.CreateAndApplyAllocation(t, flt)
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.AllocatedReplicas == 1
 	})
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(bufferSize))
 
 	// patch autoscaler to switch to relative buffer size and check if the fleet adjusts
 	_, err = patchFleetAutoscaler(fas, intstr.FromString("10%"), 1, fas.Spec.Policy.Buffer.MaxReplicas)
 	assert.Nil(t, err, "could not patch fleetautoscaler")
 
 	//10% with only one allocated GS means only one ready server
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(1))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(1))
 
 	// get the Status of the fleetautoscaler
 	fas, err = framework.AgonesClient.AutoscalingV1().FleetAutoscalers(fas.ObjectMeta.Namespace).Get(fas.Name, metav1.GetOptions{})
@@ -120,7 +120,7 @@ func TestAutoscalerBasicFunctions(t *testing.T) {
 	gp := int64(1)
 	err = stable.GameServers(defaultNs).Delete(gsa.Status.GameServerName, &metav1.DeleteOptions{GracePeriodSeconds: &gp})
 	assert.Nil(t, err)
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.AllocatedReplicas == 0 &&
 			fleet.Status.ReadyReplicas == 1 &&
 			fleet.Status.Replicas == 1
@@ -140,7 +140,7 @@ func TestAutoscalerStressCreate(t *testing.T) {
 		defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	r := rand.New(rand.NewSource(1783))
 
@@ -180,7 +180,7 @@ func TestAutoscalerStressCreate(t *testing.T) {
 					expectedReplicas = fas.Spec.Policy.Buffer.MaxReplicas
 				}
 				// the fleet autoscaler should scale the fleet now to expectedReplicas
-				framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(expectedReplicas))
+				framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(expectedReplicas))
 			} else {
 				assert.False(t, valid,
 					fmt.Sprintf("FleetAutoscaler NOT created even if the parameters are valid: %d %d %d (%s)",
@@ -270,7 +270,7 @@ func TestAutoscalerWebhook(t *testing.T) {
 		defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	fleetautoscalers := framework.AgonesClient.AutoscalingV1().FleetAutoscalers(defaultNs)
 	fas := defaultFleetAutoscaler(flt)
@@ -292,11 +292,11 @@ func TestAutoscalerWebhook(t *testing.T) {
 		assert.FailNow(t, "Failed creating autoscaler, aborting TestAutoscalerWebhook")
 	}
 	framework.CreateAndApplyAllocation(t, flt)
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.AllocatedReplicas == 1
 	})
 
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.Replicas > initialReplicasCount
 	})
 
@@ -484,7 +484,7 @@ func TestTlsWebhook(t *testing.T) {
 		defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 	}
 
-	framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 	fleetautoscalers := framework.AgonesClient.AutoscalingV1().FleetAutoscalers(defaultNs)
 	fas := defaultFleetAutoscaler(flt)
@@ -508,11 +508,11 @@ func TestTlsWebhook(t *testing.T) {
 		assert.FailNow(t, "Failed creating autoscaler, aborting TestTlsWebhook")
 	}
 	framework.CreateAndApplyAllocation(t, flt)
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.AllocatedReplicas == 1
 	})
 
-	framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+	framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 		return fleet.Status.Replicas > initialReplicasCount
 	})
 }

--- a/test/e2e/framework/framework.go
+++ b/test/e2e/framework/framework.go
@@ -132,9 +132,18 @@ func (f *Framework) WaitForGameServerState(gs *agonesv1.GameServer, state agones
 		state, gs.Namespace, gs.Name)
 }
 
-// WaitForFleetCondition waits for the Fleet to be in a specific condition or fails the test if the condition can't be met in 5 minutes.
-// nolint: dupl
-func (f *Framework) WaitForFleetCondition(t *testing.T, flt *agonesv1.Fleet, condition func(fleet *agonesv1.Fleet) bool) {
+// AssertFleetCondition waits for the Fleet to be in a specific condition or fails the test if the condition can't be met in 5 minutes.
+func (f *Framework) AssertFleetCondition(t *testing.T, flt *agonesv1.Fleet, condition func(fleet *agonesv1.Fleet) bool) {
+	t.Helper()
+	err := f.WaitForFleetCondition(t, flt, condition)
+	if err != nil {
+		// Do not call Fatalf() from go routine other than main test go routine, because it could cause a race
+		t.Fatalf("error waiting for fleet condition on fleet %v", flt.Name)
+	}
+}
+
+// WaitForFleetCondition waits for the Fleet to be in a specific condition or returns an error if the condition can't be met in 5 minutes.
+func (f *Framework) WaitForFleetCondition(t *testing.T, flt *agonesv1.Fleet, condition func(fleet *agonesv1.Fleet) bool) error {
 	t.Helper()
 	logrus.WithField("fleet", flt.Name).Info("waiting for fleet condition")
 	err := wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
@@ -147,8 +156,9 @@ func (f *Framework) WaitForFleetCondition(t *testing.T, flt *agonesv1.Fleet, con
 	})
 	if err != nil {
 		logrus.WithField("fleet", flt.Name).WithError(err).Info("error waiting for fleet condition")
-		t.Fatalf("error waiting for fleet condition on fleet %v", flt.Name)
+		return err
 	}
+	return nil
 }
 
 // WaitForFleetAutoScalerCondition waits for the FleetAutoscaler to be in a specific condition or fails the test if the condition can't be met in 2 minutes.

--- a/test/e2e/gameserverallocation_test.go
+++ b/test/e2e/gameserverallocation_test.go
@@ -51,7 +51,7 @@ func TestCreateFleetAndGameServerAllocate(t *testing.T) {
 				defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			gsa := &allocationv1.GameServerAllocation{
 				Spec: allocationv1.GameServerAllocationSpec{
@@ -88,7 +88,7 @@ func TestMultiClusterAllocationOnLocalCluster(t *testing.T) {
 				defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			// Allocation Policy #1: local cluster with desired label.
 			// This policy allocates locally on the cluster due to matching namespace with gsa and not setting AllocationEndpoints.
@@ -206,7 +206,7 @@ func TestCreateFullFleetAndCantGameServerAllocate(t *testing.T) {
 				defer fleets.Delete(flt.ObjectMeta.Name, nil) // nolint:errcheck
 			}
 
-			framework.WaitForFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+			framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
 
 			gsa := &allocationv1.GameServerAllocation{
 				Spec: allocationv1.GameServerAllocationSpec{
@@ -222,7 +222,7 @@ func TestCreateFullFleetAndCantGameServerAllocate(t *testing.T) {
 				}
 			}
 
-			framework.WaitForFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
+			framework.AssertFleetCondition(t, flt, func(fleet *agonesv1.Fleet) bool {
 				return fleet.Status.AllocatedReplicas == replicasCount
 			})
 
@@ -304,8 +304,8 @@ func TestGameServerAllocationPreferredSelection(t *testing.T) {
 		assert.FailNow(t, "could not create second fleet")
 	}
 
-	framework.WaitForFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
-	framework.WaitForFleetCondition(t, required, e2e.FleetReadyCount(required.Spec.Replicas))
+	framework.AssertFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
+	framework.AssertFleetCondition(t, required, e2e.FleetReadyCount(required.Spec.Replicas))
 
 	gsa := &allocationv1.GameServerAllocation{ObjectMeta: metav1.ObjectMeta{GenerateName: "allocation-"},
 		Spec: allocationv1.GameServerAllocationSpec{
@@ -355,7 +355,7 @@ func TestGameServerAllocationPreferredSelection(t *testing.T) {
 	assert.Nil(t, err)
 
 	// now wait for another one to come along
-	framework.WaitForFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
+	framework.AssertFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
 
 	gsa3, err := framework.AgonesClient.AllocationV1().GameServerAllocations(defaultNs).Create(gsa.DeepCopy())
 	if assert.Nil(t, err) {
@@ -399,7 +399,7 @@ func TestGameServerAllocationDuringMultipleAllocationClients(t *testing.T) {
 		assert.FailNow(t, "could not create first fleet")
 	}
 
-	framework.WaitForFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
+	framework.AssertFleetCondition(t, preferred, e2e.FleetReadyCount(preferred.Spec.Replicas))
 
 	// scale down before starting allocation
 	preferred = scaleFleetPatch(t, preferred, preferred.Spec.Replicas-20)

